### PR TITLE
Adding mlflow example

### PR DESCRIPTION
**Pull request by Hsankesara**
I have added two linear regression examples. 

* In `linear_regression_local`, all the metadata and artifacts would be stored in a local directory. 
* In `linear_regression_remote`, all the metadata and artifacts would be stored in [mlflow-server stack ](https://github.com/RADAR-base/RADAR-Docker/tree/distributed-dcompose/distributed-dcompose-stack/mlflow).
* Also, I have created `upload_data_to_postgres.py` file, to upload example dataset to local postgres databse and added instruction in README to how to do  so.

### CODE REVIEW

File: dataloader/querybuilder.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+    def get_selected_columns(self, select_cols:Union[List[str], str])-> str:
+        if type(select_cols) == str:
+            return "SELECT " + select_cols + " FROM " + self.tablename
+        elif type(select_cols) == list and all([type(col) == str for col in select_cols]):
+            return "SELECT " + ", ".join(select_cols) + " FROM " + self.tablename
+        else:
+            raise TypeError("cols must be either string of list of strings.")
+
+    def get_all_columns_groupby(self, groupby_cols:Union[List[str], str])-> str:
```
Comment (yatharthranjan) [inline dataloader/querybuilder.py:18]: Do you need the _all_columns functions here? Seem a bit redundant as the same can be achieved by passing `select_cols ='*'` in the _selected_columns functions

File: dataloader/querybuilder.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
+from typing import List, Union
+
+class QueryBuilder:
```
Comment (yatharthranjan) [inline dataloader/querybuilder.py:3]: I think this is good but not sure if everyone will prefer to use this over raw queries which might be easier and more readable instead of using function calls that mask the exact query being made.
But good to provide both options!

File: examples/linear_regression_local/README.md [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+First, install conda enviroment using:
+
+```bash
+conda env create --file examples/linear_regression_local/conda.yaml
+```
+
+Then, activate the conda enviroment using:
+
+```bash
+conda activate linear_regression_local_example
```
Comment (yatharthranjan) [inline examples/linear_regression_local/README.md:18]: What if someone is not using conda and instead using venv or something else? I think we should also provide instructions for installing directly from requirements.txt file

File: examples/linear_regression_local/train.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+
+if __name__ == "__main__":
+    warnings.filterwarnings("ignore")
+    np.random.seed(40)
+
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--alpha", default=0.1),
+    parser.add_argument("--l1-ratio", default=0.2)
+    args = parser.parse_args()
```
Comment (yatharthranjan) [inline examples/linear_regression_local/train.py:34]: might also be good to take number of iterations as an argument and use it in the model

File: examples/linear_regression_local/train.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+        print("  MAE: %s" % mae)
+        print("  R2: %s" % r2)
+
+        mlflow.log_param("alpha", alpha)
+        mlflow.log_param("l1_ratio", l1_ratio)
+        mlflow.log_metric("rmse", rmse)
+        mlflow.log_metric("r2", r2)
+        mlflow.log_metric("mae", mae)
+
+        mlflow.sklearn.log_model(lr, "model")
```
Comment (yatharthranjan) [inline examples/linear_regression_local/train.py:78]: can you name it elastic_net_wine_model

File: examples/linear_regression_remote/train.py [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+
+def eval_metrics(actual, pred):
+    rmse = np.sqrt(mean_squared_error(actual, pred))
+    mae = mean_absolute_error(actual, pred)
+    r2 = r2_score(actual, pred)
+    return rmse, mae, r2
+
+
+if __name__ == "__main__":
+    warnings.filterwarnings("ignore")
```
Comment (yatharthranjan) [inline examples/linear_regression_remote/train.py:36]: why is this used?

File: examples/linear_regression_remote/train.py [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+    dbconn = PostgresPandasWrapper(dbname="", user="", password="")
+    dbconn.connect()
+    #print(dbconn.get_response(cols=["*"], dataset="wine_dataset"))
+    query = querybuilder.get_all_columns()
+    data = dbconn.get_response(query)
+    dbconn.disconnect()
+
+    mlflow.set_tracking_uri("http://127.0.0.1:5000")
+    #Set experiment
+    mlflow.set_experiment("test-experiment")
```
Comment (yatharthranjan) [inline examples/linear_regression_remote/train.py:58]: might be better to name it more specific like `wine_elastic_net_experiment`

File: examples/upload_data_to_postgres.py [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+                        help='Password for the mention username', default="")
+    return parser
+
+def main():
+    args = get_argparser().parse_args()
+    user = args.user
+    password = args.password
+    dbname = args.dbname
+
+    df = pd.read_csv("data/wine-quality.csv")
```
Comment (yatharthranjan) [inline examples/upload_data_to_postgres.py:23]: can we make this more generic so it can upload any CSV file (with headers) into the DB? It can take the columns as header names from the CSV.

The filename to upload can be passed as an argument. Ideally, this should accept multiple files (with same header format) and upload them to the table. This will also allow us to upload the Fitbit data we have.

**Review by yatharthranjan — COMMENTED** [review]
Thanks, great work. Looking good. please find some minor comments below.

File: dataloader/querybuilder.py [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+    def __init__(self, tablename):
+        self.tablename = tablename
+
+    def get_all_columns(self) -> str:
+        return "SELECT " + "* " + "FROM " + self.tablename
+
+    def get_selected_columns(self, select_cols:Union[List[str], str])-> str:
+        if type(select_cols) == str:
+            return "SELECT " + select_cols + " FROM " + self.tablename
```
Comment (yatharthranjan) [inline dataloader/querybuilder.py:12]: BTW you can also use `f""` syntax everywhere like -
```py
return f"SELECT {select_cols} FROM {self.tablename}"
```

File: dataloader/querybuilder.py (hunk 2 above)
Comment (Hsankesara) [reply dataloader/querybuilder.py:3]: We can add raw queries directly in dataloader now. So, both the options are available.

Comment (yatharthranjan) [reply dataloader/querybuilder.py:3]: Yes I know, this was just an observation that people might prefer one or the other so its good that we have both options.  perhaps should have worded it better 😆

File: examples/linear_regression_remote/train.py (hunk 6 above)
Comment (Hsankesara) [reply examples/linear_regression_remote/train.py:36]: It was just to ignore deprecation  warnings from pandas and sklearn modules.

File: examples/upload_data_to_postgres.py (hunk 8 above)
Comment (Hsankesara) [reply examples/upload_data_to_postgres.py:23]: Sure, I can make it more generalized.

File: dataloader/querybuilder.py (hunk 9 above)
Comment (Hsankesara) [reply dataloader/querybuilder.py:12]: Thanks will do that in future.

File: dataloader/querybuilder.py (hunk 1 above)
Comment (Hsankesara) [reply dataloader/querybuilder.py:18]: Yeah, it's a bit redundant, but I added it to make it easier to get `all_columns` since not everyone who would use this module will be familiar with ``select_cols ='*'``.

File: examples/upload_data_to_postgres.py [hunk 10]
```diff
@@ -1,32 +1,47 @@
...
+        df.columns = [x.replace(" ", "_") for x in df.columns]
+
     engine = create_engine('postgresql://'+ user_info + '@localhost:5432/' + dbname)
-    df.columns = ['fixed_acidity', 'volatile_acidity', 'citric_acid', 'residual_sugar',
-           'chlorides', 'free_sulfur_dioxide', 'total_sulfur_dioxide', 'density',
-           'pH', 'sulphates', 'alcohol', 'quality']
-    df.to_sql('wine_dataset', engine, index=False)
+
+    for i, df in enumerate(df_files):
+        df.to_sql(tablenames[i], engine, index=False)
```
Comment (yatharthranjan) [inline examples/upload_data_to_postgres.py:44]: please also set, `if_exists='append'` (You can also take this from args)

So we can do something like -  

`-f fitbit-day1.csv,fitbit-day2.csv,fitbit-day3.csv -t fitbit,fitbit,fitbit`

This will read the files (having same format) and put them in the same table.

File: examples/upload_data_to_postgres.py [hunk 11]
```diff
@@ -1,32 +1,47 @@
...
     dbname = args.dbname
-
-    df = pd.read_csv("data/wine-quality.csv")
+    files = args.files
+    tablenames = args.tablenames
+    if len(tablenames) != len(files):
+        raise IOError("Number of input files must be equal to number of input tablenames")
+    print(files)
+    os_file_check(files)
+    df_files = [pd.read_csv(filepath) for filepath in files]
```
Comment (yatharthranjan) [inline examples/upload_data_to_postgres.py:36]: Instead of reading all the files into the memory at once, you can also read each as it is processed in the for loop.

File: examples/linear_regression_local/train.py [hunk 12]
```diff
@@ -55,9 +56,10 @@ def eval_metrics(actual, pred):
 
     alpha = float(args.alpha)
     l1_ratio = float(args.l1_ratio)
+    num_iterations = float(args.num_iterations)
```
Comment (yatharthranjan) [inline examples/linear_regression_local/train.py:59]: I think this should be `int` as it will always be a whole number

**Review by yatharthranjan — COMMENTED** [review]
Thanks for the changes, looks good, Just a few minor comments

File: examples/upload_data_to_postgres.py (hunk 10 above)
Comment (Hsankesara) [reply examples/upload_data_to_postgres.py:44]: Yeah I think we can search for the table and if it exist, we can insert dataframe otherwise we would create table and then add data.

File: examples/linear_regression_local/train.py (hunk 12 above)
Comment (Hsankesara) [reply examples/linear_regression_local/train.py:59]: My bad, will change it.

File: examples/upload_data_to_postgres.py (hunk 10 above)
Comment (yatharthranjan) [reply examples/upload_data_to_postgres.py:44]: I think setting the `if_exists='append'` does exactly that.

**Comment by Hsankesara** [comment]
Hey @yatharthranjan, thanks for the review. I have made the mentioned changes. Let me know if anything else is needed.

**Comment by Hsankesara** [comment]
Hey @yatharthranjan, thanks for the review. I have made the mentioned changes. Let me know if anything else is needed.

**Review by yatharthranjan — APPROVED** [review]
Thanks, LGTM
